keps/672-serial-job-execution/README.md

```diff
@@ -0,0 +1,512 @@
+- Support any other JobSet status other than Succeeded and Ready.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user, I want to fine-tune LLM using multi-node training with JobSet. I have the first
+ReplicatedJob for pre-trained model and dataset initialization, and the second ReplicatedJob for
+distributed fine-tuning.
+
+The example of JobSet looks as follows:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: fine-tune-llm
+spec:
+  startupPolicy:
+    startupPolicyOrder: InOrder
+    rules:
+      - targetReplicatedJobs:
+          - initializer
+        waitForReplicatedJobsStatus: Succeeded
+  replicatedJobs:
+    - name: initializer
+      template:
+        spec:
+          template:
+            spec:
+              containers:
+                - name: model-initializer
+                  image: docker.io/kubeflow/model-initializer
+                  volumeMounts:
+                    - mountPath: /workspace/pre-trained-model
+                      name: model-initializer
+                - name: dataset-initializer
+                  image: docker.io/kubeflow/dataset-initializer
+                  volumeMounts:
+                    - mountPath: /workspace/dataset
+                      name: dataset-initializer
+              volumes:
+                - name: dataset-initializer
+                  persistentVolumeClaim:
+                    claimName: dataset-initializer
+                - name: model-initializer
+                  persistentVolumeClaim:
+                    claimName: model-initializer
+    - name: trainer-node
+      template:
+        spec:
+          parallelism: 3
+          completions: 3
+          template:
+            spec:
+              containers:
+                - name: trainer
+                  image: docker.io/kubeflow/trainer
+                  resources:
+                    limits:
+                      nvidia.com/gpu: 5
+                  volumeMounts:
+                    - mountPath: /workspace/dataset
+                      name: dataset-initializer
+                    - mountPath: /workspace/pre-trained-model
+                      name: model-initializer
+              volumes:
+                - name: dataset-initializer
+                  persistentVolumeClaim:
+                    claimName: dataset-initializer
+                - name: model-initializer
+                  persistentVolumeClaim:
+                    claimName: model-initializer
+```
+
+#### Story 2
+
+As a user, I want to fine-tune my LLM using MPI and DeepSpeed. I have the first
+ReplicatedJob for pre-trained model and dataset initialization, the second ReplicatedJob for MPI launch
+and the third ReplicatedJob for distributed fine-tuning.
+
+The example of JobSet looks as follows:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: deepspeed-mpi
+spec:
+  startupPolicy:
+    startupPolicyOrder: InOrder
+    rules:
```

Comment:
this is going to be specific to InOrder, so what about calling `InOrderStartupRules`

I also suggest to be more explicit here in the way the user describes the rules:

```
InOrderStartupRules
  replicatedJobsWaitingOnSucceededStatus
  - initializer
  replicatedJobsWaitingOnReadyStatus
  - launcer 
```

I am not too sure about the exact name of the `replicatedJobsWaitingOnSucceededStatus|replicatedJobsWaitingOnReadyStatus` fields, please suggest others

Reply:
I like this idea, what do we think about this name: `replicatedJobsWaitForSucceededStatus` @danielvegamyhre @kannon92 @tenzen-y @ahg-g @mimowo ?

Reply:
sgtm

Reply:
> this is going to be specific to InOrder, so what about calling `InOrderStartupRules`
> 
> I also suggest to be more explicit here in the way the user describes the rules:
> 
> ```
> InOrderStartupRules
>   replicatedJobsWaitingOnSucceededStatus
>   - initializer
>   replicatedJobsWaitingOnReadyStatus
>   - launcer 
> ```
> 
> I am not too sure about the exact name of the `replicatedJobsWaitingOnSucceededStatus|replicatedJobsWaitingOnReadyStatus` fields, please suggest others

I like this, it also simplifies the API imo.

>I like this idea, what do we think about this name: replicatedJobsWaitForSucceededStatus @danielvegamyhre @kannon92 @tenzen-y @ahg-g @mimowo ?

Personally I prefer `replicatedJobsWaitingOnSucceededStatus` since it is close to a real, grammatically correct phrase, so interpreting the meaning/purpose of the variable is effortless. 

`replicatedJobsWaitForSucceededStatus` is a bit jarring / confusing to me.

Reply:
Sure, I am ok with that name too.